// @animatereactnative/accordion/lib/module/index.js
"use strict";

import { createContext, useContext, useState } from 'react';
import { Pressable } from 'react-native';
import Animated, { FadeIn, FadeOut, LinearTransition, useAnimatedStyle, useDerivedValue, withSpring } from 'react-native-reanimated';
import { jsx as _jsx } from "react/jsx-runtime";
const AnimatedPressable = Animated.createAnimatedComponent(Pressable);
const AccordionContext = /*#__PURE__*/createContext({
  isOpen: false,
  accordionIsOpen: () => {}
});

// type AccordionProvider = ContextType<typeof AccordionContext>;

// Internal Hook
const useAccordion = () => {
  const context = useContext(AccordionContext);
  if (!context) {
    throw new Error('useAccordion must be used within a Accordion.Provider');
  }
  return context;
};
function Provider({
  children,
  isOpen: initialIsOpen,
  onChange,
  style,
  ...rest
}) {
  const [isOpen, accordionIsOpen] = useState(!!initialIsOpen);
  return /*#__PURE__*/_jsx(AccordionContext.Provider, {
    value: {
      isOpen,
      accordionIsOpen: () => {
        onChange?.(!isOpen);
        accordionIsOpen(prev => {
          return !prev;
        });
      }
    },
    children: /*#__PURE__*/_jsx(Animated.View, {
      style: [style, {
        overflow: 'hidden'
      }],
      layout: LinearTransition.springify().damping(80).stiffness(200),
      ...rest,
      children: children
    })
  });
}
function Header({
  children
}) {
  const {
    accordionIsOpen
  } = useAccordion();
  return /*#__PURE__*/_jsx(AnimatedPressable, {
    onPress: () => accordionIsOpen(),
    children: children
  });
}
function Icon({
  children,
  rotation = 'clockwise'
}) {
  const {
    isOpen
  } = useAccordion();
  const rotate = useDerivedValue(() => {
    return withSpring(isOpen ? (rotation === 'clockwise' ? -1 : 1) * 180 : 0, {
      damping: 80,
      stiffness: 200
    });
  }, [isOpen, rotation]);
  const stylez = useAnimatedStyle(() => {
    return {
      transform: [{
        rotate: `${rotate.value}deg`
      }]
    };
  });
  return /*#__PURE__*/_jsx(Animated.View, {
    style: stylez,
    layout: LinearTransition.springify().damping(80).stiffness(200),
    children: children
  });
}
function Expanded({
  children,
  ...rest
}) {
  const {
    isOpen
  } = useAccordion();
  if (!isOpen) {
    return;
  }
  return /*#__PURE__*/_jsx(Content, {
    ...rest,
    children: children
  });
}
function Collapsed({
  children,
  ...rest
}) {
  const {
    isOpen
  } = useAccordion();
  if (isOpen) {
    return;
  }
  return /*#__PURE__*/_jsx(Content, {
    ...rest,
    children: children
  });
}
function Always({
  children,
  ...rest
}) {
  return /*#__PURE__*/_jsx(Content, {
    ...rest,
    children: children
  });
}
function Content({
  children,
  style,
  ...rest
}) {
  return /*#__PURE__*/_jsx(Animated.View, {
    entering: FadeIn.springify().damping(80).stiffness(200),
    exiting: FadeOut.springify().damping(80).stiffness(200),
    layout: LinearTransition.springify().damping(80).stiffness(200),
    style: [style, {
      overflow: 'hidden'
    }],
    ...rest,
    children: children
  });
}
function Sibling({
  children,
  ...rest
}) {
  return /*#__PURE__*/_jsx(Animated.View, {
    layout: LinearTransition.springify().damping(80).stiffness(200),
    ...rest,
    children: children
  });
}
export const Accordion = {
  /**
   * The main component that will handle the state of the accordion.
   *
   * @param isOpen boolean
   * @param onChange (value: boolean) => void
   */
  Accordion: Provider,
  /**
   * The header of the accordion.
   */
  Header,
  /**
   * The component that will wrap any children and it will apply a rotation to it.
   *
   * @param children
   * @param rotation clockwise | counter-clockwise
   */
  HeaderIcon: Icon,
  /**
   * This is the content that will be displayed when the accordion is open
   */
  Expanded,
  /**
   * This is the content that will be displayed when the accordion is closed
   */
  Collapsed,
  /**
   * This is the content that will always be displayed
   */
  Always,
  /**
   *
   * This is a component that will add the layout transition to any
   * sibling components. Useful when you are rendering other components
   * that are not direct children of the Accordion component.
   */
  Sibling
};
export default Accordion;
